File: priority-queues/python/heap.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Heap(object):
	def __init__(self):
		self.keys = [None]

	def insert(self, key):
		self.keys.append(key)
		self.swim(len(self))

	def delete_max(self):
		key = self.keys[1]
		self.swap(1, -1)
		del self.keys[-1]
		self.sink(1)
		return key

	def max(self):
		return self.keys[1]

	def __len__(self):
		return len(self.keys) - 1

	def empty(self):
		return len(self) == 0

	def swap(self, i, j):
		self.keys[i], self.keys[j] = self.keys[j], self.keys[i]

	def parent(self, key):
		return key//2 if key > 1 else key

	def left(self, key):
		return 2 * key

	def right(self, key):
		return 2 * key + 1

	def swim(self, key):
		parent = self.parent(key)
		while self.keys[parent] < self.keys[key]:
			self.swap(parent, key)
			key = parent
			parent = self.parent(key)

	def sink(self, key):
		while key < len(self):
			child = left = self.left(key)
			if left > len(self):
				break
			right = self.right(key)
			if right <= len(self):
				child = max([left, right], key=lambda k: self.keys[k])
			if self.keys[child] <= self.keys[key]:
				break
			self.swap(child, key)
			key = child

def main():
	heap = Heap()
	heap.insert(5)
	heap.insert(-1)
	heap.insert(7)
	heap.insert(0)
	while not heap.empty():
		print(len(heap), heap.delete_max())

if __name__ == '__main__':
	main()

File: priority-queues/java/src/Heap.java
/**
 * Created by petergoldsborough on 08/23/15.
 */

public class Heap<Key extends Comparable<Key>>
{
	public static int MINIMUM_CAPACITY = 4 + 1;

	public void insert(Key key)
	{
		_keys[++_size] = key;

		_swim(_size);

		if (_size == _keys.length) _resize();
	}

	public Key deleteMax()
	{
		Key key = _keys[1];

		_swap(1, _size--);
		_sink(1);

		if (_size == _keys.length/4) _resize();

		return key;
	}

	public Key max()
	{
		return _keys[1];
	}

	public int size()
	{
		return _size;
	}

	public boolean isEmpty()
	{
		return _size == 0;
	}

	public static void main(String[] args)
	{
		Heap<Integer> heap = new Heap<>();

		heap.insert(5);
		heap.insert(10);
		heap.insert(-2);
		heap.insert(0);

		System.out.println(heap.size());

		System.out.println(heap.deleteMax());

		System.out.println(heap.size());

		System.out.println(heap.deleteMax());
		System.out.println(heap.deleteMax());
		System.out.println(heap.deleteMax());

		System.out.println(heap.isEmpty());
	}

	private boolean _less(int first, int second)
	{
		return _keys[first].compareTo(_keys[second]) < 0;
	}

	private void _swim(int key)
	{
		int parent = _parent(key);

		while (_less(parent, key))
		{
			_swap(parent, key);

			key = parent;

			parent = _parent(key);
		}
	}

	private void _sink(int key)
	{
		while (key < _size)
		{
			int first = _child(key, 0), child = first;

			if (first >= _size) break;

			int second = _child(key, 1);

			if (second <= _size && _less(first, second)) child = second;

			if (_less(key, child)) _swap(key, child);

			else break;

			key = child;
		}
	}

	private void _swap(int first, int second)
	{
		Key temp = _keys[first];
		_keys[first] = _keys[second];
		_keys[second] = temp;
	}

	private void _resize()
	{
		int capacity = _size * 2;

		if (capacity < MINIMUM_CAPACITY) return;

		Key[] old = _keys;

		_keys = (Key[]) new Comparable[capacity];

		for (int i = 1; i <= _size; ++i)
		{
			_keys[i] = old[i];
		}

		old = null;
	}

	private int _parent(int key)
	{
		return key > 1 ? key/2 : 1;
	}

	private int _child(int key, int which)
	{
		return 2 * key + which;
	}

	private int _size = 0;

	private Key[] _keys = (Key[]) new Comparable[MINIMUM_CAPACITY];
}
